### PR TITLE
[CELEBORN-1660] Using map for workers to find worker fast

### PR DESCRIPTION
**Pull request by turboFei**
### What changes were proposed in this pull request?

Using map for workers so that we can find a worker by uniqueId fast.

### Why are the changes needed?

For large celeborn cluster, it might be slow.

- updateWorkerHeartbeatMeta
https://github.com/apache/celeborn/blob/1e77f01cd317b1dc885965d6053b391db1d42bc7/master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java#L222

- handleWorkerLost
https://github.com/apache/celeborn/blob/1e77f01cd317b1dc885965d6053b391db1d42bc7/master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala#L762-L765
### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Existing UT.

### CODE REVIEW

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java [hunk 1]
```diff
@@ -520,19 +551,19 @@ public void updatePartitionSize() {
...
-    workers.stream()
-        .filter(
-            worker ->
-                !excludedWorkers.contains(worker) && !manuallyExcludedWorkers.contains(worker))
-        .forEach(workerInfo -> workerInfo.updateDiskMaxSlots(estimatedPartitionSize));
+
+    HashSet<WorkerInfo> workers = new HashSet(workersMap.values());
+    excludedWorkers.forEach(workers::remove);
+    manuallyExcludedWorkers.forEach(workers::remove);
+    workers.forEach(workerInfo -> workerInfo.updateDiskMaxSlots(estimatedPartitionSize));
```
Comment (turboFei) [inline master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java:558]: do not look the workers

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java [hunk 2]
```diff
@@ -520,19 +551,19 @@ public void updatePartitionSize() {
...
   }
 
   public boolean isWorkerAvailable(WorkerInfo workerInfo) {
-    return !excludedWorkers.contains(workerInfo)
+    return (workerInfo.getWorkerStatus().getState() == PbWorkerStatus.State.Normal
```
Comment (turboFei) [inline master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java:562]: check state first

File: master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala [hunk 3]
```diff
@@ -1124,24 +1127,25 @@ private[celeborn] class Master(
       fileCount,
       System.currentTimeMillis(),
       requestId)
-    // unknown workers will retain in needCheckedWorkerList
-    needCheckedWorkerList.removeAll(statusSystem.workers)
+    val unknownWorkers = needCheckedWorkerList.stream()
+      .filter(w => !statusSystem.containsWorker(w))
+      .collect(Collectors.toList[WorkerInfo])
```
Comment (turboFei) [inline master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala:1132]: do not loop the workers(larger)

**Review by SteNicholas — APPROVED** [review]

File: master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala [hunk 4]
```diff
@@ -1124,24 +1127,23 @@ private[celeborn] class Master(
       fileCount,
       System.currentTimeMillis(),
       requestId)
-    // unknown workers will retain in needCheckedWorkerList
-    needCheckedWorkerList.removeAll(statusSystem.workers)
+    val unknownWorkers = needCheckedWorkerList.asScala.filterNot(statusSystem.containsWorker).asJava
```
Comment (turboFei) [inline master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala:1130]: do not loop the workers(larger)

File: master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala [hunk 5]
```diff
@@ -227,13 +227,13 @@ private[celeborn] class Master(
...
   }
-  masterSource.addGauge(MasterSource.WORKER_COUNT) { () => statusSystem.workers.size }
+  masterSource.addGauge(MasterSource.WORKER_COUNT) { () => statusSystem.getWorkers.size }
   masterSource.addGauge(MasterSource.LOST_WORKER_COUNT) { () => statusSystem.lostWorkers.size }
   masterSource.addGauge(MasterSource.EXCLUDED_WORKER_COUNT) { () =>
     statusSystem.excludedWorkers.size + statusSystem.manuallyExcludedWorkers.size
   }
   masterSource.addGauge(MasterSource.AVAILABLE_WORKER_COUNT) { () =>
-    statusSystem.workers.asScala.count { w =>
+    statusSystem.getWorkers.asScala.count { w =>
```
Comment (RexXiong) [inline master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala:236]: Why can’t we use workersMap.values directly? If we use a copy for workers, which may introduce unnecessary overhead.

Comment (RexXiong) [reply master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala:236]: It seems acceptable.

Comment (turboFei) [reply master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala:236]: UnmodifiableCollection will not copy the elements, it just limit the add/remove/putAll method access.

Comment (turboFei) [reply master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala:236]: I just set the workersMap to private to prevent that there is typo to get the worker by WorkerInfo instead of WorkerId.

Comment (RexXiong) [reply master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala:236]: > UnmodifiableCollection will not copy the elements, it just limit the add/remove/putAll method access.

I see, thanks!

**Review by RexXiong — APPROVED** [review]
LGTM

File: master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala [hunk 6]
```diff
@@ -911,27 +908,32 @@ private[celeborn] class Master(
...
-              loadAwareFetchTimeWeight,
-              requestSlots.availableStorageTypes)
-          } else {
-            SlotsAllocator.offerSlotsRoundRobin(
-              selectedWorkers,
-              requestSlots.partitionIdList,
-              requestSlots.shouldReplicate,
-              requestSlots.shouldRackAware,
-              requestSlots.availableStorageTypes)
+        statusSystem.synchronizedWorkers(new Supplier[JMap[
```
Comment (FMX) [inline master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala:911]: Why is this change necessary? Here needs to do a synchronized operation.

Comment (turboFei) [reply master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala:911]: yes，because we set the workersMap as private, so I provide a method named `synchronizedWorkers` to lock the workersMap.

File: master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala (hunk 5 above)
Comment (FMX) [reply master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala:236]: > I just set the workersMap to private to prevent that there is typo to get the worker by WorkerInfo instead of WorkerId.

All maps have this problem because get or contains will not verify its key type. Just expose the workersmap may be OK.

Comment (turboFei) [reply master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala:236]: ok

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java [hunk 7]
```diff
@@ -64,7 +66,7 @@ public abstract class AbstractMetaManager implements IMetadataHandler {
   public final Map<String, Set<Integer>> registeredAppAndShuffles =
       JavaUtils.newConcurrentHashMap();
   public final Set<String> hostnameSet = ConcurrentHashMap.newKeySet();
-  public final Set<WorkerInfo> workers = ConcurrentHashMap.newKeySet();
+  private final Map<String, WorkerInfo> workersMap = JavaUtils.newConcurrentHashMap();
```
Comment (FMX) [inline master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java:69]: Keep this field public can remove these methods below. Although there might be some incorrect usable of this map, I believe that the reviews can find out.

Comment (turboFei) [reply master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java:69]: sure

Comment (turboFei) [reply master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java]: addressed. thanks

**Review by FMX — APPROVED** [review]
LGTM. Thanks. Merged into main(v0.6.0)

**Comment by turboFei** [comment]
thanks all for the review!
